src/full/Agda/Syntax/Scope/Monad.hs

```diff
@@ -176,6 +177,38 @@ setLocalVars vars = modifyLocalVars $ const vars
 withLocalVars :: ScopeM a -> ScopeM a
 withLocalVars = bracket_ getLocalVars setLocalVars
 
+-- | Check that the newly added variable have unique names
+
+withCheckNoShadowing :: ScopeM a -> ScopeM a
+withCheckNoShadowing m = do
+  lvars0 <- getLocalVars
+  v <- m
+  lvars1 <- getLocalVars
+  checkNoShadowing lvars0 lvars1
+  pure v
+
+checkNoShadowing :: LocalVars  -- ^ Old local scope
+                 -> LocalVars  -- ^ New local scope
+                 -> ScopeM ()
+checkNoShadowing old new = do
+  -- LocalVars is currnently an AssocList so the difference between
+  -- two local scope is the left part of the new one.
+  let diff = dropEnd (length old) new
+  let nameParts = mapMaybe extractConcreteNameParts $ AssocList.keys diff
+  let conflicts = Map.filter atLeastTwo $ Map.fromListWith (++) nameParts
+  unless (Map.null conflicts) $ do
+    warning $ NicifierIssue $ ShadowingInTelescope
+            $ Map.toList conflicts
+
+  where
+
+    extractConcreteNameParts :: C.Name -> Maybe (C.Name, [Range])
```

Comment:
`nameParts` confuses me, since there is `Concrete.Name.NameParts`; but you take the whole names here.